lib/flow_builder.dart

```diff
@@ -92,23 +92,39 @@ class _FlowBuilderState<T> extends State<FlowBuilder<T>> {
   @override
   void didUpdateWidget(FlowBuilder<T> oldWidget) {
     super.didUpdateWidget(oldWidget);
-    if (oldWidget.state != widget.state) {
-      _removeListeners(dispose: oldWidget.controller == null);
-      _controller = _initController(widget.state);
-      _pages = widget.onGeneratePages(_state, List.of(_pages));
-      _history
-        ..clear()
-        ..add(_state);
-    } else if (oldWidget.controller != widget.controller) {
+
+    if (_haveStateOrControllerChanged(widget, oldWidget)) {
```

Comment:
@e200 thanks again for taking the time to refactor!

This is subjective but I personally prefer the previous implementation because it is explicit, has fewer levels of indirection, and is less verbose than the current approach. From a maintenance standpoint, I'd rather have some repeated code if it makes things easier to understand and change in the future. For example, if we want to do something different only if the state changes or only if both the state and controller have changed then we will again need to refactor this which isn't ideal imo.

@jorgecoca I'd love to hear your thoughts as well since it might just be my preference 👍

Reply:
@felangel I completely agree with you about do something when only the state has changed and have a code repeated 2 times isn’t really bad, sometimes we just need to think about code reuse if we have the same code repeated more then 2 times.

This definitely brings more complication then benefits.

Reply:
Thanks for share your opinion, it was really useful!

I’m going to close this since it is better as it is now!

Reply:
+1 to what @felangel said. It's a very minor implementation detail, but I also think the previous solution reads a bit better, even though it might have some duplication.